### ISSUE TITLE
`nnls` fails: Maximum number of iterations reached.

### ISSUE THREAD

**Issue by rguliev** [r-hyperspec/pyspc-unmix#3]
In was found in some cases that `nnls` function from `scipy.optimize` fails with "Maximum number of iterations reached." error.
However, some alternative options available that allow to get a solution.
The task is to check alternative options. Compare them and replace `nnls` with a better default solver.

**Comment by rguliev** [comment]
# Options for NNLS:

1. [scipy.optimize.nnls](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.nnls.html)
2. [scipy.optimize.lsq_linear](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.lsq_linear.html)
3. [CVXPY](https://www.cvxpy.org/)https://www.[cvxpy.org](https://www.cvxpy.org/)

# Pre-requisites for the code
[A.csv](https://github.com/user-attachments/files/17874649/A.csv)
[B.csv](https://github.com/user-attachments/files/17874650/B.csv)

# Code for comparison

```python
# %% Imports
import cvxpy as cp
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import nnls, lsq_linear
from timeit import timeit, repeat

# %% Load the data
A = np.genfromtxt('A.csv', delimiter='\t')
A = A.T

# B[0,:] and B[1,:] are two the examples where nnls fails
B = np.genfromtxt('B.csv', delimiter='\t')

# %% Helper functions
def nnls_cvxpy(A, b):
    x = cp.Variable(A.shape[1], nonneg=True)
    objective = cp.Minimize(cp.norm2(A @ x - b))
    problem = cp.Problem(objective)
    problem.solve()
    return x.value

def custom_nnls(A, X, method="nnls", **kwargs):
    
    if X.ndim == 1:
        X = X.reshape(1, -1)

    nnls_funcs = {
        "nnls": lambda x: nnls(A, x, **kwargs)[0],
        "lsq": lambda x: lsq_linear(A, x, bounds=(0, np.inf), **kwargs).x,
        "cvxpy": lambda x: nnls_cvxpy(A, x, **kwargs),
    }
    return np.apply_along_axis(nnls_funcs[method], axis=1, arr=X)

def plot_reconstrunction(i, ax=None):
    lsq = custom_nnls(A, B[i,:], method="lsq")
    cvxpy = custom_nnls(A, B[i,:], method="cvxpy")

    if ax is None:
        fig, ax = plt.subplots()
    else:
        fig = ax.get_figure()
    
    x = np.arange(B.shape[1])
    ax.plot(x, B[i,:], 'k-', label="Raw")
    ax.plot(x, (lsq @ A.T)[0,:], 'r-', label="LSQ")
    ax.plot(x, (cvxpy @ A.T)[0,:], 'b--', label="CVXPY")
    ax.legend()

    return fig, ax

# %% Check failing cases
custom_nnls(A, B[0,], method="nnls")
custom_nnls(A, B[0,], method="lsq")
custom_nnls(A, B[0,], method="cvxpy")

custom_nnls(A, B[1,], method="nnls")
custom_nnls(A, B[1,], method="lsq")
custom_nnls(A, B[1,], method="cvxpy")

# %% Visualize the results for failing cases
plot_reconstrunction(0)
plot_reconstrunction(1)

# %% Compare output with nnls when it doesn't fail
results = {
    method: custom_nnls(A, B[2:,:], method=method)
    for method in ["nnls", "lsq", "cvxpy"]
}
{k:np.round(v,4) for k,v in results.items()}
# {
# 'nnls':
#  array([[ 31.3998,   0.    ,   4.6784, 127.8892],
#         [ 27.1168,  46.3419,   5.4496,  20.3949],
#         [ 18.8738, 132.4356,   0.3378,   0.    ],
#         [105.8127,  36.696 ,  11.8655,   0.9288]]),
# 'lsq':
#  array([[ 31.3998,   0.    ,   4.6784, 127.8892],
#         [ 27.1168,  46.3419,   5.4496,  20.3949],
#         [ 18.8738, 132.4356,   0.3378,   0.    ],
#         [105.8127,  36.696 ,  11.8655,   0.9288]]),
# 'cvxpy':
#  array([[ 31.3832,   0.    ,   4.6967, 127.8873],
#         [ 27.1171,  46.3425,   5.4476,  20.3959],
#         [ 18.8739, 132.4356,   0.3376,   0.    ],
#         [105.8186,  36.6976,  11.8543,   0.9324]])
# }

np.max(np.abs(results["nnls"] - results["lsq"])) # 3.292592309911413e-10
np.max(np.abs(results["nnls"] - results["cvxpy"])) # 0.01830507027538797

# %% Benchmark performance (small)
times = {
    method: repeat(
        lambda: custom_nnls(A, B[2:,:], method=method),
        repeat=5,
        number=50
    )
    for method in ["nnls", "lsq", "cvxpy"]
}

# Print results
for method, timing in times.items():
    mean_time = np.mean(timing)
    std_time = np.std(timing)
    print(f"{method}: {mean_time:.5f}s ± {std_time:.5f}s")
# nnls:  0.06058s ± 0.01207s
# lsq:   0.22237s ± 0.01743s
# cvxpy: 1.51218s ± 0.04462s

# %% Benchmark performance (bigger dataset)
Bbig = B[2:,:].repeat(100, axis=0)
times_big = {
    method: repeat(
        lambda: custom_nnls(A, Bbig, method=method),
        repeat=5,
        number=10
    )
    for method in ["nnls", "lsq", "cvxpy"]
}

# Print results
for method, timing in times_big.items():
    mean_time = np.mean(timing)
    std_time = np.std(timing)
    print(f"{method}: {mean_time:.5f}s ± {std_time:.5f}s")
# nnls:   0.99224s ± 0.02645s
# lsq:    4.00930s ± 0.06586s
# cvxpy: 29.87123s ± 0.65334s

# %% Check reconstrunctions for the whole dataset
fig, axs = plt.subplots(2, 3, figsize=(12, 8))
for i, ax in enumerate(axs.flat):
    ax.set_title(f"Example {i}")
    plot_reconstrunction(i, ax=ax)
```

![image](https://github.com/user-attachments/assets/fa56392a-ed2e-4799-b49a-226e6d9fbc6b)

**Comment by rguliev** [comment]
Based on the comparison above it seems like switching to `lsq_linear` is a better option.

**Comment by rguliev** [comment]
Fixed in #4
